gateway-sp-comms/src/management_switch.rs

```diff
@@ -0,0 +1,592 @@
+        self.ports[port.0].local_addr()
+    }
+
+    /// Get the socket to use to communicate with an AP and the socket address
+    /// of that SP.
+    fn sp_socket(&self, port: SwitchPort) -> Option<SpSocket<'_>> {
+        // NOTE: For now, it's not possible for this method to return `None`. We
+        // control construction of `SwitchPort`s, and only hand them out for
+        // valid "ports" where we know an SP is (at least supposed to) be
+        // listening. In the future this may return `None` if there is no SP
+        // connected on this port.
+        let mut n = port.0;
+        for known_sps in [
+            &self.known_sps.switches,
+            &self.known_sps.sleds,
+            &self.known_sps.power_controllers,
+        ] {
+            if n < known_sps.len() {
+                return Some(SpSocket {
+                    socket: &self.ports[port.0],
+                    addr: known_sps[n].sp,
+                    port,
+                });
+            }
+            n -= known_sps.len();
+        }
+
+        // We only construct `SwitchPort`s with valid indices, so the only way
+        // to get here is if someone constructs two `ManagementSwitch` instances
+        // (with different port cardinalities) and then mixes up `SwitchPort`s
+        // between them (or other similarly outlandish shenanigans) - fine to
+        // panic in that case.
+        unreachable!("invalid port {:?}", port)
+    }
+}
+
+async fn recv_task<F>(inner: Arc<Inner>, mut callback: F)
+where
+    F: FnMut(SwitchPort, &'_ [u8]),
+{
+    // helper function to tag a socket's `.readable()` future with an index; we
+    // need this to make rustc happy about the types we push into
+    // `recv_all_sockets` below
+    async fn readable_with_port(
+        port: SwitchPort,
+        sock: &UdpSocket,
+    ) -> (SwitchPort, io::Result<()>) {
+        let result = sock.readable().await;
+        (port, result)
+    }
+
+    // set up collection of futures tracking readability of all switch port
+    // sockets
+    let mut recv_all_sockets = FuturesUnordered::new();
+    for (i, sock) in inner.ports.iter().enumerate() {
+        recv_all_sockets.push(readable_with_port(SwitchPort(i), sock));
+    }
+
+    let mut buf = [0; SpMessage::MAX_SIZE];
+
+    loop {
+        // `recv_all_sockets.next()` will never return `None` because we
+        // immediately push a new future into it every time we pull one out
+        // (to reregister readable interest in the corresponding socket)
+        let (port, result) = recv_all_sockets.next().await.unwrap();
+
+        // immediately push a new future requesting readability interest, as
+        // noted above
+        let sock = &inner.ports[port.0];
+        recv_all_sockets.push(readable_with_port(port, sock));
+
+        // TODO how do we handle errors here (or from `recv()` below)?
+        // currently just log them and retry, assuming any socket errors are
+        // ephemeral
```

Comment:
I thought about this perhaps too hard. Some errors would be expected (I guess) such as `std::io::ErrorKind::WouldBlock` which should be ignored. Other errors would seem indicative of some serious invariant violation. But after scrounging around in tokio a bit, it seems like this can't fail except in strange circumstances. I'd suggest that unless we **know** we can safely ignore (and log) an error we should bring panic!() and rely on SMF to restart our zone into a known consistent state.

Reply:
Ah, I like that better. I'll change this one (and the "some err other than WouldBlock" in the match below it) to panics, with a comment that we don't expect to see errors here in general.